### ISSUE TITLE
stats.kendalltau() returns invalid p-value for some inputs

### ISSUE THREAD

**Issue by thegwan** [scipy/scipy#11175]
`stats.kendalltau()` returns p-value greater than 1 for some inputs

```
import scipy.stats as stats
x = [5,2,1,3,6,4,7,8]  
y = [5,2,6,3,1,8,7,4]
stats.kendalltau(x,y,method='exact')                                                        
```

Returns `KendalltauResult(correlation=0.0, pvalue=1.0951388888888889) `

#### Scipy/Numpy/Python version information:

```
1.3.3 1.17.4 sys.version_info(major=3, minor=6, micro=8, releaselevel='final', serial=0)
```

**Comment by chrisb83** [comment]
Exact calculation was added in #8614

``
cor.test(c(5,2,1,3,6,4,7,8), c(5,2,6,3,1,8,7,4), method = "kendall", exact = TRUE)

	Kendall's rank correlation tau

data:  c(5, 2, 1, 3, 6, 4, 7, 8) and c(5, 2, 6, 3, 1, 8, 7, 4)
T = 14, p-value = 1

**Comment by WarrenWeckesser** [comment]
@chrisb83, I'm reopening the issue.  You showed the result of an R calculation, but not SciPy's result.  On the master branch, I still get a p-value of 1.095 (whether or not I used `method='exact'`):

```
In [1]: from scipy.stats import kendalltau                                      

In [2]: x = [5,2,1,3,6,4,7,8]                                                   

In [3]: y = [5,2,6,3,1,8,7,4]                                                   

In [4]: kendalltau(x, y)                                                        
Out[4]: KendalltauResult(correlation=0.0, pvalue=1.0951388888888889)

In [5]: import scipy                                                            

In [6]: scipy.__version__                                                       
Out[6]: '1.5.0.dev0+77e5b9b'
```

**Comment by chrisb83** [comment]
Sorry, I clicked the wrong button - it is a bug.

**Comment by chrisb83** [comment]
The issue is in the calculation of 'exact', if you use 'asymp', the result is ok. No argument defaults to exact for such small samples.

```
c, size = 14, 8 # These are the values that the example produces in kendalltau

new = [0.0]*(c+1)
new[0] = 1.0
new[1] = 1.0

for j in range(3,size+1):
    old = new[:]
    for k in range(1,min(j,c+1)):
        new[k] += new[k-1]
    for k in range(j,c+1):
        new[k] += new[k-1] - old[k-j]

pvalue = 2.0*sum(new)/math.factorial(size) # 1.09...
```

@Konrad0 could you take a look at your implementation if you have time? I don't have access to the reference used (Maurice G. Kendall, "Rank Correlation Methods" (4th Edition))

**Comment by Konrad0** [comment]
Thanks for spotting the bug!
Yes, I'll take a look, give me a few days as I'm rather busy at the moment.

**Comment by Konrad0** [comment]
Alright, I found the bug, thanks @chrisb83 for doing most of the detective work. Basically I made a mistake exploiting the symmetry of the array (that's the factor 2 in the p-value). Here's the corrected code:
```python
new = [0.0]*(c+1)
new[0] = 1.0
new[1] = 1.0
for j in range(3,size+1):
    old = new[:]
    for k in range(1,min(j,c+1)):
        new[k] += new[k-1]
        print(str(new[k])+', ', end='')
    for k in range(j,c+1):
        new[k] += new[k-1] - old[k-j]

if size >= 171:
    pvalue = 0.0
elif size%2 == 1:
    pvalue = 2.0*sum(new)/math.factorial(size)
else:
    pvalue = (2.0*sum(new[:-1]) + new[-1])/math.factorial(size)
```
Also, I'd like to remove the `if size >= 171:    pvalue = 0.0` and replace it by an exception or a NaN earlier in the code since it gives an incorrect result, what would you prefer? The p-value is difficult to evaluate in the exact method for large sizes, but tau may still be valid.

Let me know what you think and I'll prepare a pull request including this case as a test.

**Comment by chrisb83** [comment]
hi, sounds good, feel free to create a PR with a test case (see R benchmark above).

size >= 171:

would it work to compute the p-value as 

`exp( log(sum(new)) - log(factorial(size))` and write `log(factorial(size))` as `scipy.special.gammaln(size+1)` to avoid the overflow?

**Comment by Konrad0** [comment]
Hi everyone,

as you may have seen I submitted the pull request (#11253) fixing the issue. After a little while I noticed that my first fix was also not entirely correct and that the bug was much easier to fix than previously thought. The bug corresponds to the special case where the number of discordances is exactly half the number of total possibilities, in other words maximum "confusion", so the p-value must be 1. So a simple check before going through the whole algorithm was added, as well as several tests to make sure it works.

Also @chrisb83 I think it's a good idea to keep this separate from the issue size >= 171 and create a separate PR for that enhancement. I like your idea of using `gammaln(size+1)`, but `sum(new)` will also overflow for moderately large `size` and a p-value that isn't tiny. But we can actually do the division step by step during the iteration, then the intermediate results should stay reasonably sized, but more on that in the other PR in a few days.

**Comment by sturlamolden** [comment]
Doing the division during the iteration will accumulate tons of rounding error. We need to be smarter about this,  e.g. use a data type for which sum(new) cannot overflow.

**Comment by Konrad0** [comment]
There will definitely be roundoff error, but as of now we always get `p = 0` for `size >= 171`, so it would be at least a step in the right direction. Nonetheless, I'd be happy about a better solution @sturlamolden (keeping in mind that `sum(new)` may grow approximately like `factorial(size)`), as I consider the division while iterating far from ideal. The roundoff error could be kept minimal by replacing the only subtraction in the code by a sum, leaving us with only the rounding errors from addition and division of positive reals, although this means worse time complexity. The whole method could then be replaced by an iterated convolution, also not a solution to the rounding issue but it is a bit faster and some analytical simplifications might be possible, which I have not had time to look into.

**Comment by sturlamolden** [comment]
Well, if ```sum(new)``` grows like ```factorial(size)``` it means that the computational complexity  grows geometrically as well (and depending on implementation, so does the memory usage). I think have to accept that the exact method is only feasible for small data sets. I am therefore thinking that raising an exception is the right solution.

**Comment by chrisb83** [comment]
Closed by #11253
